pallets/investments/src/lib.rs

```diff
@@ -0,0 +1,887 @@
+
+		/// The underlying assets one can invest into
+		type AssetId: Clone + MaxEncodedLen + Parameter;
+
+		/// An incrementing per AssetId counter
+		type OrderId: AtLeast32BitUnsigned + Copy + Parameter + MaxEncodedLen;
+
+		/// Something that knows how to handle accounting for the given assets
+		/// and provides metadata about them
+		type Accountant: AssetAccountant<
+			Self::AccountId,
+			Error = DispatchError,
+			AssetId = Self::AssetId,
+			Amount = Self::Amount,
+		>;
+
+		/// A representation for an investment or redemption. Usually this
+		/// is equal to the known `Balance` type of a system.
+		type Amount: Member
+			+ Parameter
+			+ AtLeast32BitUnsigned
+			+ Default
+			+ Copy
+			+ MaxEncodedLen
+			+ FixedPointOperand
+			+ From<u64>
+			+ From<u128>
+			+ TryInto<u64>;
+
+		/// A fixed-point number which represents the value of
+		/// one currency type in terms of another.
+		type BalanceRatio: Member
+			+ Parameter
+			+ Default
+			+ Copy
+			+ FixedPointNumber<Inner = Self::Amount>;
+
+		#[pallet::constant]
+		/// The address if this pallet
+		type PalletId: Get<PalletId>;
+
+		/// The bound on how many fulfilled orders we cache until
+		/// the user needs to collect them.
+		type MaxOutstandingCollects: Get<u32>;
+
+		/// Something that can handle payments and transfers of
+		/// currencies
+		type Tokens: Mutate<Self::AccountId>
+			+ Inspect<Self::AccountId, Balance = Self::Amount>
+			+ Transfer<Self::AccountId>;
+
+		/// A possible check if investors fulfill every condition to invest into a
+		/// given asset
+		type PreConditions: PreConditions<
+			OrderType<Self::AccountId, Self::AssetId, Self::Amount>,
+			Result = bool,
+		>;
+
+		/// The weight information for this pallet extrinsics.
+		type WeightInfo: weights::WeightInfo;
+	}
+
+	#[pallet::pallet]
+	#[pallet::generate_store(pub (super) trait Store)]
+	#[pallet::without_storage_info]
+	pub struct Pallet<T>(_);
+
+	#[pallet::hooks]
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T> where
+		<T::Accountant as AssetAccountant<T::AccountId>>::AssetInfo:
+			AssetProperties<T::AccountId, Currency = CurrencyOf<T>>
+	{
+	}
+
+	#[pallet::type_value]
+	pub fn OnOrderIdEmpty<T: Config>() -> T::OrderId
+	where
+		<T::Accountant as AssetAccountant<T::AccountId>>::AssetInfo:
+			AssetProperties<T::AccountId, Currency = CurrencyOf<T>>,
+	{
+		One::one()
+	}
+
+	#[pallet::storage]
+	#[pallet::getter(fn order_id)]
+	pub type OrderId<T: Config> =
+		StorageMap<_, Blake2_128Concat, T::AssetId, T::OrderId, ValueQuery, OnOrderIdEmpty<T>>;
+
+	#[pallet::storage]
+	#[pallet::getter(fn order)]
+	pub type Orders<T: Config> = StorageDoubleMap<
+		_,
+		Blake2_128Concat,
+		T::AccountId,
+		Blake2_128Concat,
+		T::AssetId,
+		BoundedVec<Order<T::Amount, T::OrderId>, T::MaxOutstandingCollects>,
+		ValueQuery,
+	>;
+
+	#[pallet::type_value]
+	pub fn OnTotalOrderEmpty<T: Config>() -> TotalOrder<T::Amount>
```

Comment:
Why a custom value here instead of giving TotalOrder a `Default` impl for the `ValueQuery` to use?

Reply:
Good Idea!

Reply:
Good idea, done!